File: recruitexam_v2_updated/app/server/routes/proctoring.ts
import { Router } from "express";
import { supabase } from "../lib/supabase";
import { authMiddleware, type AuthRequest } from "../middleware/auth";
import { hasAiKey, verifyWebcamSnapshot } from "../lib/ai";

const router = Router();

router.use(authMiddleware);

async function verifyAndLogSnapshotViolation(
  attemptId: string,
  examId: string,
  candidateId: string,
  snapshotData: string,
  currentViolationCount: number
) {
  try {
    const analysis = await verifyWebcamSnapshot(snapshotData);
    console.log(`[AI Proctoring] Snapshot verified:`, analysis);

    // Check if candidate is currently taking an AI interview
    const { data: activeInterview } = await supabase
      .from("ai_interviews")
      .select("id")
      .eq("candidate_id", candidateId)
      .eq("status", "in_progress")
      .maybeSingle();

    const isInterview = !!activeInterview;

    let violationMessage = "";
    const prefix = isInterview ? "[AI Interview] " : "";
    if (analysis.multiple_people) {
      violationMessage = `${prefix}AI Proctoring Flag: Multiple people detected in webcam feed. (${analysis.summary})`;
    } else if (!analysis.single_person) {
      violationMessage = `${prefix}AI Proctoring Flag: No candidate face visible in webcam feed. (${analysis.summary})`;
    } else if (analysis.phone_detected) {
      violationMessage = `${prefix}AI Proctoring Flag: Secondary device / mobile phone detected. (${analysis.summary})`;
    } else if (analysis.looking_away) {
      violationMessage = `${prefix}AI Proctoring Flag: Candidate persistently looking away from screen. (${analysis.summary})`;
    }

    if (violationMessage) {
      // Find the latest override event for this attempt
      const { data: latestOverride, error: overrideErr } = await supabase
        .from("proctoring_snapshots")
        .select("captured_at")
        .eq("attempt_id", attemptId)
        .eq("event_type", "camera_check")
        .eq("message", "OVERRIDE_UNLOCK")
        .order("captured_at", { ascending: false })
        .limit(1);

      if (overrideErr) {
        console.error("[AI Proctoring] Failed to fetch latest override:", overrideErr);
      }

      // Query the database for the current violation count dynamically relative to latest override
      let countQuery = supabase
        .from("proctoring_snapshots")
        .select("id", { count: "exact", head: true })
        .eq("attempt_id", attemptId)
        .eq("event_type", "violation");

      if (latestOverride && latestOverride.length > 0) {
        countQuery = countQuery.gt("captured_at", latestOverride[0].captured_at);
      }

      const { count, error: countErr } = await countQuery;

      if (countErr) {
        console.error("[AI Proctoring] Failed to fetch existing violation count:", countErr);
      }

      const nextViolationCount = (count || 0) + 1;
      console.log(`[AI Proctoring] Security anomaly flagged. Incrementing warnings to ${nextViolationCount}.`);

      // Log the violation in the proctoring snapshots table
      await supabase.from("proctoring_snapshots").insert({
        attempt_id: attemptId,
        exam_id: examId,
        candidate_id: candidateId,
        event_type: "violation",
        violation_count: nextViolationCount,
        message: violationMessage,
        snapshot_data: snapshotData,
      });

      // Fetch recruiter ID and candidate name to trigger notification
      const { data: attemptData } = await supabase
        .from("attempts")
        .select("recruiter_id, users:candidate_id(name)")
        .eq("id", attemptId)
        .single();

      if (attemptData?.recruiter_id) {
        const candidateName = (attemptData.users as any)?.name || "Candidate";
        await supabase.from("notifications").insert({
          user_id: attemptData.recruiter_id,
          title: isInterview ? `AI Interview Security Violation - ${candidateName}` : `Security Violation Flagged - ${candidateName}`,
          body: `${violationMessage} (Warning ${nextViolationCount}/3)`,
        });
      }

      // Symmetrically trigger auto-submission if violations exceed limit of 3
      if (nextViolationCount >= 3) {
        if (isInterview) {
          console.log(`[AI Proctoring] Violation limit reached for AI Interview ${activeInterview.id}. Auto-submitting.`);
          await supabase
            .from("ai_interviews")
            .update({
              status: "completed",
              score: 0,
              summary: "AI Interview auto-submitted due to excessive proctoring violations (3/3 warnings reached).",
              feedback: "Candidate exceeded the maximum allowable security violations during the AI face-to-face round. Interview terminated.",
              submitted_at: new Date().toISOString(),
            })
            .eq("id", activeInterview.id);
        } else {
          console.log(`[AI Proctoring] Violation limit reached for attempt ${attemptId}. Auto-submitting exam.`);
          // Finalize the attempt overall score using background queue
          await supabase
            .from("attempts")
            .update({
              status: "completed",
              submitted_at: new Date().toISOString(),
            })
            .eq("id", attemptId);

          const { gradingQueue } = await import("../lib/queue");
          gradingQueue.push(attemptId);
        }
      }
    }
  } catch (err) {
    console.error("[AI Proctoring] Failed to analyze snapshot:", err);
  }
}

router.post("/events", async (req: AuthRequest, res) => {
  try {
    const { attempt_id, exam_id, event_type, violation_count, message, snapshot_data } = req.body;

    if (!attempt_id || !exam_id || !event_type) {
      res.status(400).json({ error: "attempt_id, exam_id, and event_type are required" });
      return;
    }

    const { data: attempt, error: attemptError } = await supabase
      .from("attempts")
      .select("id, candidate_id, exam_id")
      .eq("id", attempt_id)
      .single();

    if (attemptError || !attempt) {
      res.status(404).json({ error: "Attempt not found" });
      return;
    }

    if (attempt.candidate_id !== req.user!.id || attempt.exam_id !== exam_id) {
      res.status(403).json({ error: "Forbidden" });
      return;
    }

    const { data, error } = await supabase
      .from("proctoring_snapshots")
      .insert({
        attempt_id,
        exam_id,
        candidate_id: req.user!.id,
        event_type,
        violation_count: violation_count || 0,
        message: message || null,
        snapshot_data: snapshot_data || null,
      })
      .select()
      .single();

    if (error) {
      res.status(400).json({ error: error.message });
      return;
    }

    // Backend enforcement of 3-strikes rule logged from frontend
    if (violation_count >= 3) {
      if (message && message.includes("[AI Interview]")) {
        console.log(`[AI Proctoring] Frontend flagged 3 violations. Auto-submitting AI interview.`);
        await supabase
          .from("ai_interviews")
          .update({
            status: "completed",
            score: 0,
            summary: "AI Interview auto-submitted due to excessive proctoring violations (3/3 warnings reached).",
            feedback: "Candidate exceeded the maximum allowable security violations during the AI face-to-face round. Interview terminated.",
            submitted_at: new Date().toISOString(),
          })
          .eq("candidate_id", req.user!.id)
          .eq("status", "in_progress");
      } else {
        console.log(`[AI Proctoring] Frontend flagged 3 violations. Auto-submitting exam.`);
        await supabase
          .from("attempts")
          .update({
            status: "completed",
            submitted_at: new Date().toISOString(),
          })
          .eq("id", attempt_id);
      }
    }

    // Asynchronously trigger AI verification of the snapshot if snapshot data is present and AI key is configured
    if (snapshot_data && hasAiKey() && (event_type === "snapshot" || event_type === "camera_check")) {
      void verifyAndLogSnapshotViolation(attempt_id, exam_id, req.user!.id, snapshot_data, violation_count || 0);
    }

    res.json({ message: "Proctoring event logged", event: data });
  } catch (err) {
    console.error("Proctoring event error:", err);
    res.status(500).json({ error: "Server error" });
  }
});

router.get("/attempt/:attemptId", async (req: AuthRequest, res) => {
  try {
    const { attemptId } = req.params;
    const { role, id } = req.user!;

    const { data: attempt, error: attemptError } = await supabase
      .from("attempts")
      .select("id, candidate_id, recruiter_id")
      .eq("id", attemptId)
      .single();

    if (attemptError || !attempt) {
      res.status(404).json({ error: "Attempt not found" });
      return;
    }

    const canView =
      role === "admin" ||
      attempt.candidate_id === id ||
      attempt.recruiter_id === id;

    if (!canView) {
      res.status(403).json({ error: "Forbidden" });
      return;
    }

    const { data, error } = await supabase
      .from("proctoring_snapshots")
      .select("*")
      .eq("attempt_id", attemptId)
      .order("captured_at", { ascending: false });

    if (error) {
      res.status(400).json({ error: error.message });
      return;
    }

    res.json({ events: data || [] });
  } catch (err) {
    console.error("Proctoring fetch error:", err);
    res.status(500).json({ error: "Server error" });
  }
});

router.get("/exam/:examId/summary", async (req: AuthRequest, res) => {
  try {
    if (!["admin", "recruiter"].includes(req.user!.role)) {
      res.status(403).json({ error: "Forbidden" });
      return;
    }

    const { examId } = req.params;
    const collegeId = req.query.collegeId as string | undefined;

    let attemptsQuery = supabase
      .from("attempts")
      .select("id, candidate_id, recruiter_id, status, score, users:candidate_id(name, email)")
      .eq("exam_id", examId);

    if (req.user!.role === "recruiter") {
      attemptsQuery = attemptsQuery.eq("recruiter_id", req.user!.id);
    }

    if (collegeId) {
      const { data: profiles } = await supabase
        .from("candidate_profiles")
        .select("user_id")
        .eq("college_id", collegeId);
      const userIds = (profiles || []).map(p => p.user_id);
      if (userIds.length === 0) {
        res.json({ summary: [] });
        return;
      }
      attemptsQuery = attemptsQuery.in("candidate_id", userIds);
    }

    const { data: attempts, error: attemptError } = await attemptsQuery;
    if (attemptError) {
      res.status(400).json({ error: attemptError.message });
      return;
    }

    const attemptIds = (attempts || []).map((attempt) => attempt.id);
    const { data: events } = attemptIds.length
      ? await supabase
          .from("proctoring_snapshots")
          .select("attempt_id, event_type, violation_count, message, captured_at")
          .in("attempt_id", attemptIds)
      : { data: [] as any[] };

    const summary = (attempts || []).map((attempt: any) => {
      const attemptEvents = (events || []).filter((event: any) => event.attempt_id === attempt.id);
      const violations = attemptEvents.filter((event: any) => event.event_type === "violation");
      return {
        attemptId: attempt.id,
        candidateId: attempt.candidate_id,
        candidateName: attempt.users?.name || "Candidate",
        candidateEmail: attempt.users?.email || "",
        status: attempt.status,
        score: attempt.score || 0,
        snapshots: attemptEvents.filter((event: any) => event.event_type === "snapshot").length,
        violations: violations.length,
        lastViolation: violations.slice().sort((a: any, b: any) => new Date(b.captured_at).getTime() - new Date(a.captured_at).getTime())[0]?.message || "",
      };
    });

    res.json({ summary });
  } catch (err) {
    console.error("Proctoring summary error:", err);
    res.status(500).json({ error: "Server error" });
  }
});

router.get("/exam/:examId/active-monitoring", async (req: AuthRequest, res) => {
  try {
    if (!["admin", "recruiter"].includes(req.user!.role)) {
      res.status(403).json({ error: "Forbidden" });
      return;
    }

    const { examId } = req.params;
    const collegeId = req.query.collegeId as string | undefined;

    let attemptsQuery = supabase
      .from("attempts")
      .select("id, candidate_id, recruiter_id, status, started_at, users:candidate_id(name, email)")
      .eq("exam_id", examId)
      .eq("status", "in_progress");

    if (req.user!.role === "recruiter") {
      attemptsQuery = attemptsQuery.eq("recruiter_id", req.user!.id);
    }

    if (collegeId) {
      const { data: profiles } = await supabase
        .from("candidate_profiles")
        .select("user_id")
        .eq("college_id", collegeId);
      const userIds = (profiles || []).map(p => p.user_id);
      if (userIds.length === 0) {
        res.json({ attempts: [] });
        return;
      }
      attemptsQuery = attemptsQuery.in("candidate_id", userIds);
    }

    const { data: attempts, error: attemptError } = await attemptsQuery;
    if (attemptError) {
      res.status(400).json({ error: attemptError.message });
      return;
    }

    const attemptIds = (attempts || []).map((attempt) => attempt.id);
    const { data: events } = attemptIds.length
      ? await supabase
          .from("proctoring_snapshots")
          .select("attempt_id, event_type, violation_count, message, captured_at")
          .in("attempt_id", attemptIds)
          .order("captured_at", { ascending: false })
      : { data: [] as any[] };

    const activeAttempts = (attempts || []).map((attempt: any) => {
      const attemptEvents = (events || []).filter((event: any) => event.attempt_id === attempt.id);
      
      const latestEvent = attemptEvents[0];
      const isLocked = latestEvent?.event_type === "violation";

      const latestOverride = attemptEvents.find(
        (event: any) => event.event_type === "camera_check" && event.message === "OVERRIDE_UNLOCK"
      );

      let warningCount = 0;
      if (latestOverride) {
        const overrideTime = new Date(latestOverride.captured_at).getTime();
        warningCount = attemptEvents.filter(
          (event: any) => event.event_type === "violation" && new Date(event.captured_at).getTime() > overrideTime
        ).length;
      } else {
        warningCount = attemptEvents.filter((event: any) => event.event_type === "violation").length;
      }

      return {
        attemptId: attempt.id,
        candidateId: attempt.candidate_id,
        candidateName: attempt.users?.name || "Candidate",
        candidateEmail: attempt.users?.email || "",
        startedAt: attempt.started_at,
        warningCount,
        isLocked,
        latestMessage: latestEvent?.message || null,
        latestCapturedAt: latestEvent?.captured_at || null,
      };
    });

    res.json({ attempts: activeAttempts });
  } catch (err) {
    console.error("Active monitoring error:", err);
    res.status(500).json({ error: "Server error" });
  }
});

router.post("/attempt/:attemptId/override", async (req: AuthRequest, res) => {
  try {
    if (!["admin", "recruiter"].includes(req.user!.role)) {
      res.status(403).json({ error: "Forbidden" });
      return;
    }

    const { attemptId } = req.params;
    const { role, id: userId } = req.user!;

    const { data: attempt, error: attemptError } = await supabase
      .from("attempts")
      .select("id, exam_id, candidate_id, recruiter_id")
      .eq("id", attemptId)
      .single();

    if (attemptError || !attempt) {
      res.status(404).json({ error: "Attempt not found" });
      return;
    }

    if (role !== "admin" && attempt.recruiter_id !== userId) {
      res.status(403).json({ error: "Forbidden" });
      return;
    }

    const { data, error } = await supabase
      .from("proctoring_snapshots")
      .insert({
        attempt_id: attemptId,
        exam_id: attempt.exam_id,
        candidate_id: attempt.candidate_id,
        event_type: "camera_check",
        violation_count: 0,
        message: "OVERRIDE_UNLOCK",
        snapshot_data: null,
      })
      .select()
      .single();

    if (error) {
      res.status(400).json({ error: error.message });
      return;
    }

    res.json({ message: "Override unlock logged", event: data });
  } catch (err) {
    console.error("Override unlock error:", err);
    res.status(500).json({ error: "Server error" });
  }
});

export default router;

File: server_py/app/proctoring.py
import datetime
from typing import Any, Dict, List, Literal, Optional

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel, Field

from .auth_router import get_current_user, require_roles
from .audit import record_audit_event
from .db import db
from .rate_limit import limiter

router = APIRouter(prefix="/api/proctoring", tags=["proctoring"])


class ProctoringEventRequest(BaseModel):
    attempt_id: str = Field(min_length=1, max_length=128)
    exam_id: str = Field(min_length=1, max_length=128)
    event_type: Literal["camera_check", "snapshot", "violation", "submission"]
    message: Optional[str] = Field(default=None, max_length=2_000)
    snapshot_data: Optional[str] = Field(default=None, max_length=1_500_000)
    typing_speed_wpm: Optional[int] = Field(default=0, ge=0, le=500)


class OverrideSnapshotRequest(BaseModel):
    violation_severity: str


def _severity_for_event(event_type: str, violation_count: int, explicit: Optional[str] = None) -> str:
    if explicit in {"low", "medium", "high", "critical"}:
        return explicit
    if event_type != "violation":
        return "low"
    if violation_count >= 3:
        return "critical"
    if violation_count == 2:
        return "high"
    return "medium"


def _relative_time(started_at: Optional[str], captured_at: Optional[str]) -> str:
    if not started_at or not captured_at:
        return "00:00"
    try:
        start = datetime.datetime.fromisoformat(str(started_at).replace("Z", "+00:00"))
        captured = datetime.datetime.fromisoformat(str(captured_at).replace("Z", "+00:00"))
        seconds = max(0, int((captured - start).total_seconds()))
        return f"{seconds // 60:02d}:{seconds % 60:02d}"
    except Exception:
        return "00:00"


def _latest_event(events: List[Dict[str, Any]]) -> Optional[Dict[str, Any]]:
    return events[0] if events else None


async def _assert_attempt_access(attempt_id: str, user: Dict[str, Any]) -> Dict[str, Any]:
    attempt_res = await db.from_("attempts").select(
        "*, users:candidate_id(name, email), exams:exam_id(title)"
    ).eq("id", attempt_id).maybeSingle()
    attempt = attempt_res.data
    if not attempt:
        raise HTTPException(status_code=404, detail="Attempt not found")

    role = user.get("role")
    if role == "candidate" and attempt.get("candidate_id") != user["id"]:
        raise HTTPException(status_code=403, detail="Forbidden")
    if role == "recruiter" and attempt.get("recruiter_id") != user["id"]:
        raise HTTPException(status_code=403, detail="Forbidden")
    if role not in {"candidate", "recruiter", "admin"}:
        raise HTTPException(status_code=403, detail="Forbidden")
    return attempt


@router.post("/events")
@limiter.limit("120/minute")
async def log_event(
    request: Request,
    req: ProctoringEventRequest,
    user: Dict[str, Any] = Depends(get_current_user),
):
    if user.get("role") != "candidate":
        raise HTTPException(status_code=403, detail="Only candidates can submit proctoring events")
    attempt = await _assert_attempt_access(req.attempt_id, user)
    if attempt.get("exam_id") != req.exam_id:
        raise HTTPException(status_code=400, detail="Attempt does not belong to this exam")

    if req.message == "OVERRIDE_UNLOCK":
        raise HTTPException(status_code=400, detail="Reserved proctoring event message")

    prior_violations = await db.from_("proctoring_snapshots").select("id", count="exact", head=True).eq(
        "attempt_id", req.attempt_id
    ).eq("event_type", "violation")
    violation_count = (prior_violations.count or 0) + 1 if req.event_type == "violation" else 0
    severity = _severity_for_event(req.event_type, violation_count)
    payload = {
        "attempt_id": req.attempt_id,
        "exam_id": req.exam_id,
        "candidate_id": attempt["candidate_id"],
        "event_type": req.event_type,
        "violation_count": violation_count,
        "violation_severity": severity,
        "message": req.message,
        "snapshot_data": req.snapshot_data,
        "typing_speed_wpm": int(req.typing_speed_wpm or 0),
    }

    res = await db.from_("proctoring_snapshots").insert(payload).select().single()
    if res.error:
        raise HTTPException(status_code=400, detail=res.error.message)
    return {"event": res.data}


@router.get("/attempt/{attemptId}")
async def get_attempt_events(attemptId: str, user: Dict[str, Any] = Depends(get_current_user)):
    await _assert_attempt_access(attemptId, user)
    events_res = await db.from_("proctoring_snapshots").select("*").eq("attempt_id", attemptId).order("captured_at", ascending=False)
    return {"events": events_res.data or []}


@router.get("/attempts/{attemptId}/timeline")
async def get_attempt_timeline(attemptId: str, user: Dict[str, Any] = Depends(get_current_user)):
    attempt = await _assert_attempt_access(attemptId, user)
    events_res = await db.from_("proctoring_snapshots").select("*").eq("attempt_id", attemptId).order("captured_at", ascending=True)
    timeline = []
    for event in events_res.data or []:
        timeline.append({
            **event,
            "relativeTime": _relative_time(attempt.get("started_at"), event.get("captured_at")),
        })
    return {"timeline": timeline}


@router.get("/exam/{examId}/summary")
async def get_exam_summary(
    examId: str,
    collegeId: Optional[str] = None,
    user: Dict[str, Any] = Depends(require_roles(["recruiter", "admin"])),
):
    query = db.from_("attempts").select(
        "*, users:candidate_id(name, email), exams:exam_id(title)"
    ).eq("exam_id", examId)
    if user["role"] == "recruiter":
        query = query.eq("recruiter_id", user["id"])
    attempts = query.order("started_at", ascending=False)
    attempts_res = await attempts
    attempt_rows = attempts_res.data or []

    if collegeId:
        profiles_res = await db.from_("candidate_profiles").select("user_id").eq("college_id", collegeId)
        allowed = {row["user_id"] for row in profiles_res.data or []}
        attempt_rows = [attempt for attempt in attempt_rows if attempt.get("candidate_id") in allowed]

    summary = []
    for attempt in attempt_rows:
        events_res = await db.from_("proctoring_snapshots").select("*").eq("attempt_id", attempt["id"]).order("captured_at", ascending=False)
        events = events_res.data or []
        violations = [event for event in events if event.get("event_type") == "violation"]
        latest_violation = violations[0] if violations else None
        candidate = attempt.get("users") or {}
        summary.append({
            "attemptId": attempt["id"],
            "candidateId": attempt.get("candidate_id"),
            "candidateName": candidate.get("name") or "Candidate",
            "candidateEmail": candidate.get("email") or "",
            "status": attempt.get("status"),
            "snapshots": len(events),
            "violations": len(violations),
            "lastViolation": latest_violation.get("message") if latest_violation else None,
            "lastViolationAt": latest_violation.get("captured_at") if latest_violation else None,
        })
    return {"summary": summary}


@router.get("/exam/{examId}/active-monitoring")
async def get_active_monitoring(
    examId: str,
    collegeId: Optional[str] = None,
    user: Dict[str, Any] = Depends(require_roles(["recruiter", "admin"])),
):
    query = db.from_("attempts").select(
        "*, users:candidate_id(name, email)"
    ).eq("exam_id", examId).eq("status", "in_progress")
    if user["role"] == "recruiter":
        query = query.eq("recruiter_id", user["id"])
    attempts_res = await query.order("started_at", ascending=False)
    attempt_rows = attempts_res.data or []

    if collegeId:
        profiles_res = await db.from_("candidate_profiles").select("user_id").eq("college_id", collegeId)
        allowed = {row["user_id"] for row in profiles_res.data or []}
        attempt_rows = [attempt for attempt in attempt_rows if attempt.get("candidate_id") in allowed]

    active = []
    for attempt in attempt_rows:
        events_res = await db.from_("proctoring_snapshots").select("*").eq("attempt_id", attempt["id"]).order("captured_at", ascending=False)
        events = events_res.data or []
        latest = _latest_event(events)
        latest_override = next((event for event in events if event.get("event_type") == "admin_override"), None)
        violations = [event for event in events if event.get("event_type") == "violation"]
        latest_violation = violations[0] if violations else None
        is_unlocked = latest_override and latest_violation and latest_override.get("captured_at") >= latest_violation.get("captured_at")
        warning_count = 0 if is_unlocked else max([int(event.get("violation_count") or 1) for event in violations], default=0)
        candidate = attempt.get("users") or {}
        active.append({
            "attemptId": attempt["id"],
            "candidateId": attempt.get("candidate_id"),
            "candidateName": candidate.get("name") or "Candidate",
            "candidateEmail": candidate.get("email") or "",
            "startedAt": attempt.get("started_at"),
            "warningCount": warning_count,
            "isLocked": bool(latest and latest.get("event_type") == "violation" and not is_unlocked),
            "lastEvent": latest.get("message") if latest else None,
        })
    return {"attempts": active}


@router.post("/attempt/{attemptId}/override")
async def override_attempt(attemptId: str, user: Dict[str, Any] = Depends(require_roles(["recruiter", "admin"]))):
    attempt = await _assert_attempt_access(attemptId, user)
    payload = {
        "attempt_id": attemptId,
        "exam_id": attempt["exam_id"],
        "candidate_id": attempt["candidate_id"],
        "event_type": "admin_override",
        "violation_count": 0,
        "violation_severity": "low",
        "message": "Attempt unlocked by an authorized reviewer",
    }
    res = await db.from_("proctoring_snapshots").insert(payload).select().single()
    if res.error:
        raise HTTPException(status_code=400, detail=res.error.message)
    await record_audit_event(actor_id=user["id"], action="PROCTORING_ATTEMPT_OVERRIDE", resource="attempt", resource_id=attemptId, payload={"candidate_id": attempt["candidate_id"]})
    return {"success": True, "event": res.data}


@router.post("/snapshots/{snapshotId}/override")
async def override_snapshot(
    snapshotId: str,
    req: OverrideSnapshotRequest,
    user: Dict[str, Any] = Depends(require_roles(["recruiter", "admin"])),
):
    if req.violation_severity not in {"low", "medium", "high", "critical"}:
        raise HTTPException(status_code=400, detail="Invalid severity")

    snapshot_res = await db.from_("proctoring_snapshots").select("*").eq("id", snapshotId).maybeSingle()
    snapshot = snapshot_res.data
    if not snapshot:
        raise HTTPException(status_code=404, detail="Snapshot not found")
    await _assert_attempt_access(snapshot["attempt_id"], user)

    update_res = await db.from_("proctoring_snapshots").update({
        "violation_severity": req.violation_severity,
    }).eq("id", snapshotId).select().single()
    if update_res.error:
        raise HTTPException(status_code=400, detail=update_res.error.message)
    await record_audit_event(actor_id=user["id"], action="PROCTORING_SNAPSHOT_OVERRIDE", resource="proctoring_snapshot", resource_id=snapshotId, payload={"violation_severity": req.violation_severity})
    return {"snapshot": update_res.data}
